### PR TITLE
SONARJAVA-3741 Rule S6202: Operator "instanceof" should be used inste…

### PR DESCRIPTION
**Pull request by margarita-nedzelska-sonarsource**
…ad of "A.class.isInstance()"

### CODE REVIEW

**Comment by margarita-nedzelska-sonarsource** [comment]
As far as this method is present since Java 1.1, this check could be applied to all the versions of Java.

File: java-checks/src/main/resources/org/sonar/l10n/java/rules/java/S6202_java.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "title": "Operator \"instanceof\" should be used instead of \"A.class.isInstance()\"",
+  "type": "CODE_SMELL",
+  "status": "ready",
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "tags": [
+    "java8"
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/resources/org/sonar/l10n/java/rules/java/S6202_java.json:10]: You should update the rule metadata to remove this tag.

File: java-checks/src/main/java/org/sonar/java/checks/IsInstanceMethodCheck.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+    if (IS_INSTANCE_MATCHER.matches(methodInvocationTree)) {
+      ExpressionTree methodSelect = methodInvocationTree.methodSelect();
+      if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+        ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
+        getClassIdentifier(expression)
+          .ifPresent(identifierTree -> reportIssue(tree, String.format(MESSAGE, identifierTree.name(), identifierTree.name())));
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/IsInstanceMethodCheck.java:58]: You can use `Replace this usage of \"%1$s.class.isInstance()\" with \"instanceof %1$s\"` and provide only one time the identifier name to the format method.

File: java-checks/src/main/java/org/sonar/java/checks/IsInstanceMethodCheck.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S6202")
+public class IsInstanceMethodCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Replace this usage of \"%s.class.isInstance()\" with \"instanceof %s\"";
```
Comment (quentin-jaquier-sonarsource) [inline java-checks/src/main/java/org/sonar/java/checks/IsInstanceMethodCheck.java:38]: Don't forget the dot at the end of the issue message:
```suggestion
  private static final String MESSAGE = "Replace this usage of \"%s.class.isInstance()\" with \"instanceof %s\".";
```

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**